### PR TITLE
Rubocop Configuration - Phase 2

### LINKED ISSUE #42: Rubocop Configuration - Phase 2
# Dependencies

- Don't start until #39 (Phase 1) is done

# Phase 2

- [ ] Relevant Cops for _RSpec_ are enabled explicitly for
  - [ ] [`-performance`](https://github.com/rubocop/rubocop-performance)
  - [ ] [`-rake`](https://github.com/rubocop/rubocop-rake)
  - [ ] [`-slim`](https://github.com/r7kamura/rubocop-slim)
  - [ ] [`-capybara`](https://github.com/rubocop/rubocop-capybara)
  - [ ] [`-thread_safety`](https://github.com/rubocop/rubocop-thread_safety)
  - [ ] Add up to `base`:
    - [ ] [`-bundler`](https://docs.rubocop.org/rubocop/1.50/cops_bundler.html)
    - [ ] [`-layout`](https://docs.rubocop.org/rubocop/1.50/cops_layout.html)
    - [ ] [`-lint`](https://docs.rubocop.org/rubocop/1.50/cops_lint.html)
    - [ ] [`-metrics`](https://docs.rubocop.org/rubocop/1.50/cops_metrics.html) (to consider)
    - [ ] [`-naming`](https://docs.rubocop.org/rubocop/1.50/cops_naming.html) (to consider)
    - [ ] [`-security`](https://docs.rubocop.org/rubocop/1.50/cops_security.html)
- [ ] _development-environment_ is, if needed, adjusted to use this configuration properly
- [ ] Following projects are referencing _Shimmer_'s rubocop and are stripped down of customized cops and have PRs open (waiting for this one) where the new styling is applied.
  - [ ] KS Core
  - [ ] Recordsale (??)
  - [ ] Flip Vinyl
  - [ ] Anymator
  - [ ] Copa Coupona ([PR](https://github.com/nerdgeschoss/copa-coupona/pull/364))
  - [ ] Tiny Farms
  - [ ] eaze
  - [ ] pull-request
  - [ ] Screamshot
  - [ ] app
  - [ ] accounting
  - [ ] LAIC (?)
- [ ] Document in _Shimmer_ how to base a project on the _Rubocop_ configuration.

### PR DESCRIPTION
**Pull request by Adeynack**
fixes #42

### CODE REVIEW

**Comment by Adeynack** [comment]
# Performance/ChainArrayAllocation

After benchmarking, I decided to NOT enable that cop, because it forces a coding style that feels backwards and in my opinion less clear (multiple lines vs chained-functional-calls), while there actually is an even more memory-efficient alternative.

In short, the cop is forcing this code

```ruby
a.map(&:to_i).map { _1 % 3 == 0 ? _1 : nil }.compact.take(100)
```

to be written this way instead

```ruby
  a.map!(&:to_i)
  a.map! { _1 % 3 == 0 ? _1 : nil }
  a.compact!
  a.take(100)
```

in order to be more performant.

And more performant it is. Here is a benchmark I wrote that compares them (code at the end of this message). We can see that what the cop suggests allocates `20 240` times less then the pure functional way.

```
Comparison:
   performance style:       1054 allocated
    functional style:   21333767 allocated - 20240.77x more
```

However, this is because we are "holding it wrong" most of the times, and because the implementation in Ruby is kind of evil, probably for backwards compatibility reasons.

TL;DR: Use `.lazy`.

In other languages, those return "iterators" automatically, making the whole thing "lazy". There are no `.map` creating a whole new array, and `.filter` creating yet another one. Those return all iterators and each element flows to the end of the chain before things are used, either as iterator, or "re-assembled" in an array or set or hash or whatever is needed.

Now, if we simply start the chain with `.lazy`, we end up using even LESS memory – 3 times less in fact – while keeping a more readable and concise coding style.

```
Comparison:
lazy functional style:       1832 allocated
   performance style:       5694 allocated - 3.11x more
    functional style:   21335307 allocated - 11645.91x more
```

Full code of the benchmark (using GEM `benchmark-memory`):

```ruby
# frozen_string_literal: true

source_a = 1_000_000.times.map(&:to_s).to_a

def functional_style(a)
  a.map(&:to_i).map { _1 % 3 == 0 ? _1 : nil }.compact
end

def functional_lazy_style(a)
  functional_style(a.lazy)
end

def performance_style(a)
  a.map!(&:to_i)
  a.map! { _1 % 3 == 0 ? _1 : nil }
  a.compact!
end

Benchmark.memory do |x|
  x.report("functional style") { functional_style(source_a) }
  x.report("performance style") { performance_style(source_a) }
  x.report("lazy functional style") { functional_lazy_style(source_a) }

  x.compare!
end
```
